### PR TITLE
add align_corners and align_mode for image_resize

### LINKED ISSUE #15276: resize_nearest与tensorflow.image.resize_nearest_neighbor差异
两个接口的定义分别为
```
paddle.fluid.layers.resize_nearest(input, out_shape=None, scale=None, name=None, actual_shape=None)
```

```
tensorflow.image.resize_nearest_neighbor( images, size, align_corners=False, name=None)
```
测试中发现，同样的输入，只有在tensorflow接口里align_corners=True的前提下，输出才无diff

请问是否有方法，在align_corners=False的情况下，使PaddlePaddle和Tensorflow得到同样的输出结果呢

### PR DESCRIPTION
**Pull request by tink2123**
fix #15276

### CODE REVIEW

File: paddle/fluid/operators/interpolate_op.cc [hunk 1]
```diff
@@ -82,6 +82,18 @@ class InterpolateOpMaker : public framework::OpProtoAndCheckerMaker {
...
+        "align_corners",
+        "an optinal bool. Defaults to True. "
+        "If True, the centers of 4 corner pixels of the input and output "
+        "tensors are aligned, preserving the values at the corner pixels, "
+        "if Flase, are not aligned")
+        .SetDefault(true);
+    AddAttr<int>("align_mode",
+                 "(int, default \'0\'), align_corners mode , can be \'0\' "
+                 "for pytorch calculation method, can be \'1\' for "
+                 "tensorflow calculation method.")
```
Comment (heavengate) [inline paddle/fluid/operators/interpolate_op.cc]: not to write pytorch and tf here, explain the index calucation by fomula

File: paddle/fluid/operators/interpolate_op.cc [hunk 2]
```diff
@@ -98,6 +110,67 @@ class InterpolateOpMaker : public framework::OpProtoAndCheckerMaker {
           to perform linear interpolation first in one direction, and then 
           again in the other direction.
 
+          Align_corners and align_mode are optinal parameters,The calculation method 
+          of interpolation can be selected by them.
+          
+          Example:
+
+          for scale:
```
Comment (heavengate) [inline paddle/fluid/operators/interpolate_op.cc]: for -> For

File: paddle/fluid/operators/interpolate_op.cc [hunk 3]
```diff
@@ -98,6 +110,67 @@ class InterpolateOpMaker : public framework::OpProtoAndCheckerMaker {
...
+              
+              input : (N,C,H_in,W_in)
+              output: (N,C,H_out,W_out) where:
+              
+              H_out = (H_{in}+0.5) * scale_{factor} - 0.5
+              W_out = (W_{in}+0.5) * scale_{factor} - 0.5
+
+
+          case 2:
+              align_corners = False , align_mode = 1
```
Comment (heavengate) [inline paddle/fluid/operators/interpolate_op.cc]: case 1/2 here should be if/else ？

File: paddle/fluid/operators/interpolate_op.cu [hunk 4]
```diff
@@ -91,15 +102,23 @@ __global__ void KeBilinearInterpFw(
     int channel_id = out_id_w / out_img_size;
 
     int out_img_idy = (out_id_w % out_img_size) / out_img_w;
-    int in_img_idy = ratio_h * out_img_idy;
+    int in_img_idy = (align_mode == 0 && !align_corners)
+                         ? static_cast<int>(ratio_h * (out_img_idy + 0.5) - 0.5)
+                         : static_cast<int>(ratio_h * out_img_idy);
```
Comment (heavengate) [inline paddle/fluid/operators/interpolate_op.cu:108]: add 
```suggestion
                         : static_cast<int>(ratio_h * out_img_idy);
      in_img_idy = max(0, in_img_idy) 
```
also in other calculations like ratio_h * (out_img_idy + 0.5) - 0.5

File: python/paddle/fluid/tests/unittests/test_bilinear_interp_op.py [hunk 5]
```diff
@@ -151,6 +190,8 @@ def init_test_case(self):
         self.out_h = 64
         self.out_w = 32
         self.out_size = np.array([66, 40]).astype("int32")
+        self.align_corners = False
+        self.align_mode = 0
```
Comment (heavengate) [inline python/paddle/fluid/tests/unittests/test_bilinear_interp_op.py]: former code of blinear interpolate is base in align_corners = True? Set align_corners = True in existed tests and add align_corner = False new tests

**Review by heavengate — COMMENTED** [review]
add "test=develop" in your commit message to trigger CI test

File: paddle/fluid/operators/interpolate_op.cc (hunk 1 above)
Comment (tink2123) [reply paddle/fluid/operators/interpolate_op.cc]: fixed

File: paddle/fluid/operators/interpolate_op.cc (hunk 2 above)
Comment (tink2123) [reply paddle/fluid/operators/interpolate_op.cc]: fixed

File: paddle/fluid/operators/interpolate_op.cc (hunk 3 above)
Comment (tink2123) [reply paddle/fluid/operators/interpolate_op.cc]: done

File: paddle/fluid/operators/interpolate_op.cu (hunk 4 above)
Comment (tink2123) [reply paddle/fluid/operators/interpolate_op.cu:108]: done

File: python/paddle/fluid/tests/unittests/test_bilinear_interp_op.py (hunk 5 above)
Comment (tink2123) [reply python/paddle/fluid/tests/unittests/test_bilinear_interp_op.py]: done

File: paddle/fluid/operators/interpolate_op.cc [hunk 6]
```diff
@@ -82,6 +82,18 @@ class InterpolateOpMaker : public framework::OpProtoAndCheckerMaker {
...
+        "align_corners",
+        "an optinal bool. Defaults to True. "
+        "If True, the centers of 4 corner pixels of the input and output "
+        "tensors are aligned, preserving the values at the corner pixels, "
+        "if Flase, are not aligned")
+        .SetDefault(true);
+    AddAttr<int>("align_mode",
+                 "(int, default \'1\'), can be \'0\' for "
+                 "src_idx = scale*(dst_indx+0.5)-0.5 , can be \'1\' for "
+                 "src_idx = scale*dst_index .")
```
Comment (heavengate) [inline paddle/fluid/operators/interpolate_op.cc]: add "in bilinear interpolation"

File: paddle/fluid/operators/interpolate_op.cu [hunk 7]
```diff
@@ -91,15 +102,25 @@ __global__ void KeBilinearInterpFw(
...
     int h_id = (in_img_idy < in_img_h - 1) ? 1 : 0;
-    T h1lambda = ratio_h * out_img_idy - in_img_idy;
+    T h1lambda = (align_mode == 0 && !align_corners)
+                     ? ratio_h * (out_img_idy + 0.5) - 0.5 - in_img_idy
+                     : ratio_h * out_img_idy - in_img_idy;
     T h2lambda = 1.f - h1lambda;
 
     int out_img_idx = tid % out_img_w;
-    int in_img_idx = ratio_w * out_img_idx;
+    int in_img_idx = (align_mode == 0 && !align_corners)
```
Comment (heavengate) [inline paddle/fluid/operators/interpolate_op.cu]: add blank after |

File: paddle/fluid/operators/interpolate_op.cc (hunk 6 above)
Comment (heavengate) [reply paddle/fluid/operators/interpolate_op.cc]: add "optional for bilinear interpolation"

File: paddle/fluid/operators/interpolate_op.cc [hunk 8]
```diff
@@ -82,6 +82,18 @@ class InterpolateOpMaker : public framework::OpProtoAndCheckerMaker {
...
+        .SetDefault(1);
```
Comment (heavengate) [inline paddle/fluid/operators/interpolate_op.cc]: align mode default as 0

File: paddle/fluid/operators/interpolate_op.cc [hunk 9]
```diff
@@ -98,6 +110,64 @@ class InterpolateOpMaker : public framework::OpProtoAndCheckerMaker {
           to perform linear interpolation first in one direction, and then 
           again in the other direction.
 
+          Align_corners and align_mode are optinal parameters,The calculation method
```
Comment (qingqing01) [inline paddle/fluid/operators/interpolate_op.cc]: ,The -> , the

File: paddle/fluid/operators/interpolate_op.h [hunk 10]
```diff
@@ -48,20 +51,31 @@ template <typename T>
...
-                                  const int c, const int out_h,
-                                  const int out_w) {
+                                  const int c, const int out_h, const int out_w,
+                                  const bool align_corners,
+                                  const bool align_mode) {
   auto input_t = EigenTensor<T, 4>::From(input);
   auto output_t = EigenTensor<T, 4>::From(*output);
   for (int k = 0; k < out_h; k++) {  // loop for images
-    int y_n = static_cast<int>(ratio_h * k);
+    int y_n = (align_mode == 0 && !align_corners)
```
Comment (qingqing01) [inline paddle/fluid/operators/interpolate_op.h]: Before line 59 (for loop):

```c++
bool align_flag = (align_mode == 0 && !align_corners);
```

Then  use `align_flag` in following codes:

```
int y_n = align_flag ? xx : xx;
```

File: paddle/fluid/operators/interpolate_op.h [hunk 11]
```diff
@@ -106,19 +121,31 @@ static void BilinearInterpolationGrad(const Tensor& output_grad,
...
                                       const int in_w, const int n, const int c,
-                                      const int out_h, const int out_w) {
+                                      const int out_h, const int out_w,
+                                      const bool align_corners,
+                                      const int align_mode) {
   auto input_grad_t = EigenTensor<T, 4>::From(*input_grad);
   auto output_grad_t = EigenTensor<T, 4>::From(output_grad);
   for (int k = 0; k < out_h; k++) {  // loop for images
-    int y_n = static_cast<int>(ratio_h * k);
+    int y_n = (align_mode == 0 && !align_corners)
```
Comment (qingqing01) [inline paddle/fluid/operators/interpolate_op.h]: same as above.

File: python/paddle/fluid/layers/nn.py [hunk 12]
```diff
@@ -6527,7 +6527,9 @@ def image_resize(input,
                  scale=None,
                  name=None,
                  resample='BILINEAR',
-                 actual_shape=None):
+                 actual_shape=None,
+                 align_corners=True,
+                 align_mode=1):
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py:6532]: Please make sure the default value compatible with older versions

File: paddle/fluid/operators/interpolate_op.h (hunk 10 above)
Comment (tink2123) [reply paddle/fluid/operators/interpolate_op.h]: done

File: python/paddle/fluid/layers/nn.py (hunk 12 above)
Comment (tink2123) [reply python/paddle/fluid/layers/nn.py:6532]: Confirmed, the default values are compatible

**Review by heavengate — DISMISSED** [review]
LGTM

**Review by qingqing01 — DISMISSED** [review]

**Review by XiaoguangHu01 — DISMISSED** [review]
同意合入

**Review by heavengate — APPROVED** [review]
LGTM

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**